APE21.rst

```diff
@@ -0,0 +1,170 @@
+  minor releases, but be pragmatic and occasionally allow such changes if they would
+  not affect many users.
+* Add deprecation warnings or future warnings in minor and major releases but only
+  go through with the breaking changes in major releases (either the next major
+  release following when a warning was added, or subsequent ones if the changes
+  are deemed to be high impact enough to wait longer).
+
+With this in place, we would effectively be striving to maximize stability over
+two-year periods but users would see deprecation warnings to prepare them for
+changes. This is in contrast to the current LTS system where users might not see
+a deprecation warning until the next LTS release as deprecations are not typically
+backported.
+
+Guidelines on deprecations and API changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The main proposed guidelines are as follows:
+
+* API changes should only be carried out in major releases.
+* Any API changes should be preceded by deprecation warnings that should be
+  visible to users in at least one minor release cycle prior to the API change
+  being carried out.
+* New deprecations should not be introduced in bugfix releases - this will
+  ensure that deprecations will therefore be emitted for at least 6 months
+  before a major release.
+* If developers wish to make an API change at a point in time where the next release
+  is a major release, they should introduce the deprecation in the major release
+  and carry out the change in the following major release.
+
+As with all guidelines, there are exceptions where we can deviate from these guidelines:
+
+* API changes can be carried out in major releases without deprecation if it is
+  not possible to have a deprecation phase due to the nature of the change.
+* Changes breaking or changing the behavior of code may be needed in order to
+  fix bugs. Such changes can be made in bugfix or minor releases and do not need
+  to be considered API changes.
+* Changes to exception types or warning/error messages may occasionally be
+  needed in non-major releases.
+
+These exceptions will require judgment calls on the part of maintainers, but any
+such changes should be minimized as much as possible and clearly communicated to
+users (see `Communication with users`_).
+
+Communication with users
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+It is imperative that we communicate effectively with users so that they understand
+what to expect from different releases:
+
+* We should include documentation for the core package that describes the
+  release policy above, namely that in general minor releases will not
+  break/remove functionality, but that major releases might.
+
+* We should ensure that any changes in releases which deviate from the
+  guidelines above are clearly communicated to users - for example any API/breaking
+  changes in minor releases should be included in the "What's New" page for the
+  minor release.
```

Comment:
Yes to this, but this to me makes even more clear that we really don't want messages like "the exception text for so-and-so has been reworded to ..." (nor should we start adding that to the changelog). This has to be *useful* information to people who are worried about results changing.

Reply:
Sounds good!

Reply:
I disagree for reasons stated above. In practice this doesn't (or shouldn't) happen so frequently that it will generate too much noise, and I specifically think that there should be a barrier to making changes like this.

Especially the changelog should be complete and any exception/warning message format change should be noted as an API change. There should not be any hidden changes in releases, esp. ones that *might* break downstream code/tests.

The What's New perhaps doesn't need this however.

Reply:
I guess I disagree: I don't think there should be a barrier to making spelling corrections, textual clarifications, etc. It should be fine to do those with edit-on-github, without also having to create a changelog entry.

Conceptually, I see our API as being described in docstrings, and those (if complete) tell which type of error is raised, not what message the error contains. (Similarly, what type of object is returned, not its `repr`.)

Anyway, don't want to make more of this than it is worth. Most important is what we can agree on - that these types of API changes would not have to be in What's New.